library/pkcs7_write.c

```diff
@@ -0,0 +1,829 @@
+        {
+            // bytes in step = new currently written bytes - old currently written
+            // bytes
+            bytes_written_in_step = *size - ( *ptr - *start ) - curr_len;
+            rc = pkcs7_write_data( start, size, ptr, WRITE_TAG, &tag,
+                              bytes_written_in_step, 0 );
+        }
+        if( rc ) 
+        	mbedtls_printf( "ERROR: Failed to add algorithm ID to PKCS7\n" );
+    }
+
+    return ( rc );
+}
+
+static int pkcs7_set_version( unsigned char **start, size_t *size, unsigned char **ptr,
+                      mbedtls_pkcs7_info *pkcs7_info )
+{
+    int rc;
+    int version = 1;
+    // for now only version 1
+    rc = pkcs7_set_algo_id( start, size, ptr, pkcs7_info );
+    if( !rc ) 
+    {
+        rc = pkcs7_write_data( start, size, ptr, WRITE_VERSION, &version,
+                          sizeof( version ), 0 );
+        if( rc ) 
+        	mbedtls_printf( "ERROR: Failed to add version to PKCS7\n" );
+    }
+
+    return ( rc );
+}
+
+static int pkcs7_set_signed_data( unsigned char **start, size_t *size,
+                         unsigned char **ptr, mbedtls_pkcs7_info *pkcs7_info )
+{
+    int rc;
+    size_t curr_len;
+    unsigned char tag = MBEDTLS_ASN1_CONSTRUCTED | MBEDTLS_ASN1_SEQUENCE;
+    rc = pkcs7_set_version( start, size, ptr, pkcs7_info );
+    if( !rc ) 
+    {
+        curr_len = *size - ( *ptr - *start );
+        rc = pkcs7_write_data( start, size, ptr,
+                          WRITE_TAG, &tag,
+                          curr_len, 0 );
+        if( rc )
+            mbedtls_printf(
+                "ERROR: Failed to add signed data's SEQUENCE header to PKCS7\n");
+    }
+
+    return ( rc );
+}
+
+static int pkcs7_set_pkcs7_oid( unsigned char **start, size_t *size, unsigned char **ptr,
+                       mbedtls_pkcs7_info *pkcs7_info )
+{
+    int rc;
+    size_t curr_len;
+    unsigned char tag = MBEDTLS_ASN1_CONTEXT_SPECIFIC | MBEDTLS_ASN1_CONSTRUCTED;
+    rc = pkcs7_set_signed_data( start, size, ptr, pkcs7_info );
+
+    if( !rc ) 
+    {
+        curr_len = *size - ( *ptr - *start );
+        rc = pkcs7_write_data( start, size, ptr,
+                          WRITE_TAG,
+                          &tag, curr_len, 0 );
+        if( !rc ) 
+        {
+            curr_len = *size - ( *ptr - *start );
+            rc = pkcs7_write_data(
+                     start, size, ptr, WRITE_OID,
+                     (void *) MBEDTLS_OID_PKCS7_SIGNED_DATA,
+                     strlen( MBEDTLS_OID_PKCS7_SIGNED_DATA ), curr_len );
+        }
+        if( rc )
+            mbedtls_printf(
+                "ERROR: Failed to add PKCS7 OID/SEQUENCE header to PKCS7\n" );
+    }
+    return ( rc );
+}
+
+static int pkcs7_start_generation( unsigned char **start, size_t *size, 
+					unsigned char **ptr, mbedtls_pkcs7_info *pkcs7_info )
+{
+	//This will call every other function before anything is actually written
+	//rip stack memory but writing backwards gives me a headache
```

Comment:
Unfortunately I'm not sure there's any other way of writing DER-encoded ASN.1, as we need to know the length of contents before we can write the length of its container...

Regarding stack memory, at first glance it looks like all `pkcs7_set_xxx()` functions are static and called only once, so I expect any compiler to inline all of them (even with `-Os`/`-Oz`), and the result to use no more stack than if it were a single very long function.

FWIW, I find the way you wrote it quite elegant.

Reply:
Thank you for you're kind words. As a developer who is just getting started and is easily confused, this means a lot! Also, I appreciate you explaining all of this, I am enjoying learning (for once).